src/main/java/org/reaktivity/nukleus/amqp/internal/stream/AmqpServerFactory.java

```diff
@@ -0,0 +1,613 @@
+
+                if (bufferSlot != BufferPool.NO_SLOT)
+                {
+                    MutableDirectBuffer decodeBuffer = bufferPool.buffer(bufferSlot);
+                    decodeBuffer.putBytes(bufferSlotOffset, buffer, offset, length);
+                    bufferSlotOffset += length;
+                    buffer = decodeBuffer;
+                    offset = 0;
+                    length = bufferSlotOffset;
+                }
+
+                boolean decoderStateChanged = true;
+                while (length > 0 && decoderStateChanged)
+                {
+                    int consumed;
+                    DecoderState previous = decodeState;
+                    consumed = decodeState.decode(buffer, offset, length);
+                    decoderStateChanged = previous != decodeState;
+                    offset += consumed;
+                    length -= consumed;
+                }
+
+                if (length > 0)
+                {
+                    if (bufferSlot == BufferPool.NO_SLOT)
+                    {
+                        bufferSlot = bufferPool.acquire(initialId);
+                    }
+                    MutableDirectBuffer decodeBuffer = bufferPool.buffer(bufferSlot);
+                    decodeBuffer.putBytes(0, buffer, offset, length);
+                    bufferSlotOffset = length;
+                }
+                else
+                {
+                    releaseBufferSlotIfNecessary();
+                }
+            }
+        }
+
+        private void releaseBufferSlotIfNecessary()
+        {
+            if (bufferSlot != BufferPool.NO_SLOT)
+            {
+                bufferPool.release(bufferSlot);
+                bufferSlot = BufferPool.NO_SLOT;
+                bufferSlotOffset = 0;
+            }
+        }
+
+        private void onEnd(
+            EndFW end)
+        {
+            // TODO
+            releaseBufferSlotIfNecessary();
+        }
+
+        private void onAbort(
+            AbortFW abort)
+        {
+            // TODO
+            releaseBufferSlotIfNecessary();
+        }
+
+        private void onWindow(
+            WindowFW window)
+        {
+            final int replyCredit = window.credit();
+
+            replyBudget += replyCredit;
+            replyPadding += window.padding();
+
+            final int initialCredit = bufferPool.slotCapacity() - initialBudget;
+            doWindow(supplyTraceId.getAsLong(), initialCredit);
+        }
+
+        private void onReset(
+            ResetFW reset)
+        {
+            // TODO
+        }
+
+        private void onSignal(
+            SignalFW signal)
+        {
+            // TODO
+        }
+
+        private void onAmqpHeader(
+            AmqpProtocolHeaderFW header)
+        {
+            if (header != null)
+            {
+                if (isAmqpHeaderValid(header))
+                {
+                    doAmqpHeader(header.buffer(), header.offset(), header.limit());
```

Comment:
Suggest switching this to receive a major version, a minor version and a revision version, then construct the AMQP protocol header in `doAmqpHeader` rather than echoing back the same `header` as was received.

You'll certainly need an approach that doesn't rely on the input when handling the invalid protocol header case.